### PR TITLE
boards: shields: nrf2240ek: 500mA pmic current limit on bootup

### PR DESCRIPTION
**Pull request by ankuns**
Bringing objects close to the PCB antenna of the nRF2240EK might
cause voltage drops which could in effect reset the nRF2240 what is
invisible to the remaining application.
To counteract this phenomenon the current limit of the PMIC
present on the nRF2240EK is increased to 500mA.

### CODE REVIEW

File: boards/shields/nrf2240ek/src/shield.c [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+	};
+
+	return i2c_write_dt(spec, tx_buf, 3);
+}
+
+static int nrf2240ek_pmic_init(void)
+{
+	int err;
+	struct i2c_dt_spec bus = I2C_DT_SPEC_GET(DT_NODELABEL(nrf2240ek_pmic));
+
```
Comment (ankuns) [inline boards/shields/nrf2240ek/src/shield.c:51]: Rationale for direct i2c writes.

The `regulator.h` API promises the function `regulator_set_current_limit`
https://github.com/nrfconnect/sdk-zephyr/blob/fb6ca20a06ba31a3dc1b86d6986d48ad4a5760b0/include/zephyr/drivers/regulator.h#L604

 but it is not implemented
https://github.com/nrfconnect/sdk-zephyr/blob/fb6ca20a06ba31a3dc1b86d6986d48ad4a5760b0/drivers/regulator/regulator_npm1300.c#L637

File: boards/shields/nrf2240ek/src/shield.c [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+	return 0;
+}
+
+static int shield_init(void)
+{
+	return nrf2240ek_pmic_init();
+}
+
+SYS_INIT(shield_init, POST_KERNEL, CONFIG_I2C_INIT_PRIORITY);
```
Comment (anangl) [inline boards/shields/nrf2240ek/src/shield.c:76]: This does not guarantee that the I2C driver will be initialized first. The priority value should be greater.

File: west.yml [hunk 3]
```diff
@@ -72,7 +72,7 @@ manifest:
     # https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/zephyr/guides/modules.html
     - name: zephyr
       repo-path: sdk-zephyr
-      revision: fb6ca20a06ba31a3dc1b86d6986d48ad4a5760b0
```
Comment (WalkingTalkingPotato) [inline west.yml:75]: Is this related to the rest of the changes?

File: boards/shields/nrf2240ek/src/shield.c (hunk 1 above)
Comment (WalkingTalkingPotato) [reply boards/shields/nrf2240ek/src/shield.c:51]: `regulator_set_current_limit` for VBUS is not implemented because USB interface of PMIC is not considered a regulator. BUCK1, BUCK2, LDO1 and LDO2 are regulators, and they indeed don't support setting of a current limit.

There is `charger` interface you could use, I'll post a more detailed answer below in a moment.

File: west.yml (hunk 3 above)
Comment (ankuns) [reply west.yml:75]: To setup current limit on bootup I need to put C code within shield. It was impossible in current sdk-zephyr, but in upstream zephyr there is a commit that makes it possible. Cherry-pick from upstream zephyr and sha update is in separate commit. It could be done in a separate PR, but if I did it in a separate PR one could ask in that PR why is this cherry-pick necessary.
So this change is necessary here, but not core of the change.

**Review by WalkingTalkingPotato — CHANGES_REQUESTED** [review]
I believe you could implement the workaround you need using existing functions. The driver with the interface you need is "npm1300 charger":

https://github.com/nrfconnect/sdk-zephyr/blob/fb6ca20a06ba31a3dc1b86d6986d48ad4a5760b0/dts/bindings/sensor/nordic%2Cnpm1300-charger.yaml
https://github.com/nrfconnect/sdk-zephyr/blob/fb6ca20a06ba31a3dc1b86d6986d48ad4a5760b0/drivers/sensor/nordic/npm1300_charger/npm1300_charger.c

Most of its settings come from device tree and it implements `sensor` interface.

I believe the device tree setting `vbus-limit-microamp` is what you would want to use, see example:
https://github.com/nrfconnect/sdk-zephyr/blob/fb6ca20a06ba31a3dc1b86d6986d48ad4a5760b0/boards/shields/npm1300_ek/npm1300_ek.overlay#L52

It sets a different register than the one you have used. I believe the difference is that the one you used is intended for usage for when USB host tells the device (in the process of USB enumeration if I'm not mistaken) what is the actual maximal current it supports, and the one that is set by this dts setting is intended for default initial value that is used before host tells what is the limit. nPM1300 does not handle this conversation with a host, that's why it is implemented this way.

If you're sure that you want to use the registers you've used, you'll have to call `sensor_attr_set()`:

```
const struct sensor_value currentLimit = { .val1 = 0, val2 = 500000 };
sensor_attr_set(charger_device, SENSOR_CHAN_CURRENT, SENSOR_ATTR_CONFIGURATION, &currentLimit);
```

Please note that you must be sure that the source is indeed able to provide the current you need. I believe that in some cases just drawing 500 mA would be an issue.

File: boards/shields/nrf2240ek/src/shield.c (hunk 2 above)
Comment (ankuns) [reply boards/shields/nrf2240ek/src/shield.c]: addrssed in commit https://github.com/nrfconnect/sdk-nrf/pull/17140/commits/ece601ac28dd41077ba63a1f899ca720a68f00d4

**Comment by ankuns** [comment]
Regarding https://github.com/nrfconnect/sdk-nrf/pull/17140#pullrequestreview-2279810179 https://github.com/nrfconnect/sdk-nrf/pull/17140#discussion_r1743507393  https://github.com/nrfconnect/sdk-nrf/pull/17140#discussion_r1743344024 and taking into account the F2F discussion, would it be okay to merge it as is.
It is mentioned that this is a workaround, it is intentionally not mentioned what type of pmic is in use.

**Comment by ankuns** [comment]
@anangl @WalkingTalkingPotato I changed the approach. The workaround uses now the advised `npm1300_charger`. No more direct i2c writes. Please review.

**Review by anangl — APPROVED** [review]

**Review by WalkingTalkingPotato — APPROVED** [review]
